_episodes/07-b-numpy.md

```diff
@@ -0,0 +1,280 @@
+print(r1_array)
+~~~
+{: .language-python}
+
+~~~
+[0.0, 0.1, 0.0]
+[0.  0.1 0. ]
+~~~
+{: .output}
+
+However, one way numpy arrays and lists vary significantly is that you can easily perform element-wise operations on numpy arrays without loops. **This is advantageous because loops in Python are slow. However, `numpy` functions are optimized to run quickly.** If possible, you should avoid loops when working with numpy arrays.
+
+You can add two arrays together, multiply arrays by scalars, or do element-wise multiplcation of arrays.
+
+For example, you can multiply two numpy arrays to get their element-wise product. This means that given two vectors `a = np.array([a0, a1, a2])` and `b = np.array([b0, b1, b2])`, `a * b = [a0*b0, a1*a1, a2*b2]`.
+
+In contrast, if `a` and `b` were lists, you would get an error. 
+
+> ## Check your understanding
+> What does the following code result in?
+>
+> ~~~
+> a1 = np.array([2, 1, 0])
+> a2 = np.array([1, 3, 5])
+> 
+> print(a1 * a2)
+> ~~~
+> {: .language-python}
+> 
+> What happens if a1 and a2 are lists?
+>
+>> ## Answer
+>> 
+>> The code block results in 
+>> ~~~
+>> [2 3 0]
+>> ~~~
+>> {: .output}
+>> The first element is `a1[0]*a2[0]`, the second element is `a1[1]*a2[1]`, and the third element is `a1[2]*a2[2]`.
+>>
+>> If a1 and a2 were lists instead, you would get a `TypeError`. `TypeError: can't multiply sequence by non-int of type 'list'`.
+>>
+> {: .solution}
+{: .challenge}
+
+## Returning to Tabular Data
+
+Let's consider our tabular data.
+
+~~~
+import os
+import numpy
+
+distance_file = os.path.join('data', 'distance_data_headers.csv')
+distances = numpy.genfromtxt(fname=distance_file, delimiter=',', dtype='unicode')
+headers = distances[0]
+data = distances[1:]
+data = data.astype(numpy.float)
+~~~
+{: .language-python}
+
+In a previous lesson, we calculated the mean of each column of the array using the `range` function and a `for` loop. This was good because it reminded us of the `range` function and how `for` loops worked. However, the `numpy.mean` function will let us do that without a for loop.
+
+Checkout the [numpy documentation](https://docs.scipy.org/doc/numpy/reference/generated/numpy.mean.html) for the mean function. Can you figure out what argument we might change to get the mean of each column?
+
+We will use the `axis` argument. If we do not specify `axis`, the mean of the entire array is computed. For the `axis` argument, rows correspond to axis 0, and columns correspond to axis 1. This is similar to slices where you give the indices as `row, column`.
+
+~~~
+column_averages = np.mean(data, axis=0)
+print(column_averages)
+~~~
+{: .language-python}
+
+~~~
+[5000.5          10.87695093    7.34234496   11.20979133   10.9934435 ]
+~~~
+{: .output}
+
+> ## Check your understanding
+> How would you use the `mean` function and slicing to only get the mean of the columns of data (ie, exluding the frame). Save your answer
+> as a variable called `data_averages`.
+>> ## Answer
```

Comment:
I don't think most students could actually figure out what to do here based on what they have learned so far.  I would be more explicit about the solution in the question.  Within your mean function, take a slice of the data such that you only get the mean of the columns of data, i.e. exclude the frame numbers column.  Save your answer as a variable called data_averages.